audio: igo_nr: added igo_nr component

fuyuntsuo opened the pull request:
Added intelliGo noise reduction wrapper codes and topology.

The proprietary static library shall be released by intelliGo via private channel upon request.

cujomalainey commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,523 @@
+	igo_nr_get_igo_params(dev);
+
+	cd->igo_lib_config.algo_name = "igo_nr";
+	cd->igo_lib_config.in_ch_num = 1;
```
please don't implicitly assume you are getting one channel, please check the current params against what you support and reject anything not valid so we dont run into corrupted data

fuyuntsuo replied:
These are used to configure library as well. Currently the library only support single channel because mono processing takes 250MCPS already.

cujomalainey replied:
yes, that is fine, but SOF can try and open the component with an invalid format/channels/rate, even if your topology specifies certain config, do not assume that the next configuration engineer will make the same wise decisions. The code should defend against invalid configs.

fuyuntsuo replied:
Replied above. Now the wrapper would reject anything other than sample rate = 16000 and s16/s32 in `set_capture_func`.

cujomalainey commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,523 @@
+				struct comp_buffer, source_list);
+
+	/* Clear in/out buffers */
+	memset(cd->in, 0, IGO_NR_IN_BUF_LENGTH * sizeof(int16_t));
```
again please make generic to format.

fuyuntsuo replied:
`cd->in` is the 16-bit working buffer for `IgoLibProcess`. For example, 32-bit data are truncated in `igo_nr_capture_s32` before buffering in `cd->in`.

cujomalainey commented on src/include/ipc/topology.h:
```diff
@@ -47,6 +47,7 @@ enum sof_comp_type {
 	SOF_COMP_ASRC,		/**< Asynchronous sample rate converter */
 	SOF_COMP_DCBLOCK,
 	SOF_COMP_SMART_AMP,		/**< smart amplifier component */
+	SOF_COMP_IGO_NR,        /**< intelliGo NR component */
```
enums are legacy, your uuid is enough, please remove this

fuyuntsuo replied:
okay. done.

cujomalainey commented on src/include/kernel/abi.h:
```diff
@@ -29,7 +29,7 @@
 
 /** \brief SOF ABI version major, minor and patch numbers */
 #define SOF_ABI_MAJOR 3
-#define SOF_ABI_MINOR 18
```
without enum change you don't need to rev the abi either

fuyuntsuo replied:
okay. thanks!

cujomalainey commented on src/include/sof/audio/igo_nr/igo_nr_comp.h:
```diff
@@ -0,0 +1,48 @@
+#define IGO_MEM_SIZE         (40 * 1024)
+
+/* IGO_NR component private data */
+struct igo_nr_comp_data {
```
do you need this anywhere else other than your c file? if not, we should move it there to make sure it is truly private

fuyuntsuo replied:
`IGO_MEM_SIZE`, or more specifically, `uint8_t igo_heap[IGO_MEM_SIZE];` is a workaround because we failed to allocated 40KB memory on heap. The heap on SOF seems far less than 40KB.

cujomalainey replied:
not igo memsize, im referring to `igo_nr_comp_data` also please rename to just `comp_data` that way it will error out in the compiler if it leaks scope into another component

fuyuntsuo replied:
Ah thanks. Fixed now. We took tdfb as reference and didn't know there is a such checking mechanism.

cujomalainey commented on src/include/sof/audio/igo_nr/igo_nr_comp.h:
```diff
@@ -0,0 +1,48 @@
+#if defined __XCC__ && XCHAL_HAVE_HIFI3 == 1
+#include <xtensa/config/core-isa.h>
+#else
+#error "No XCC/HIFI3 found. Cannot build IGO_NR module."
```
I think we have a kconfig HiFi, please just use that for the dependency on HiFi

fuyuntsuo replied:
It must be Hifi3 of TGL, which is only available during compiling. I've checked in Kconfig and there is no such flag.

cujomalainey replied:
The wrapper should not have a hard dependency on platform, you should be able to reuse the interface and recompile your library for other platforms.

lgirdwood replied:
In general the wrapper should be able to build on any platform as it's generic C. This means you only need to port your closed source binary for other architecture and the wrapper will just work.

fuyuntsuo replied:
Agree with you about the definition of wrapper. I will change accordingly.

However if the toolchain to compile static library doesn't match the one to compile SOF framework, it might just show a plain linking error. Any suggestion to deal with this situation.

fuyuntsuo replied:
Another point to note is that igo_nr static library might have various HiFi version but never gcc, because the MCPS will be too large without Xtensa hardware acceleration.

lgirdwood replied:
I would put some text in the Kconfig help stating what ISA the binary supports today and with some link to your company. This way users can contact your company and request your product with a different ISA config etc. e.g. "This feature links to a proprietary binary X that currently is supported on xtensa ISA config Y. Please see URL for any questions about the binary".

fuyuntsuo replied:
sounds good. thanks for the suggestion and I will update the Kconfig description.

cujomalainey commented on src/include/sof/audio/igo_nr/igo_nr_comp.h:
```diff
@@ -0,0 +1,48 @@
+	struct IgoStreamData igo_stream_data_out;
+	struct comp_data_blob_handler *model_handler;
+	struct sof_igo_nr_config config;	    /**< blob data buffer */
+	int16_t in[IGO_NR_IN_BUF_LENGTH];	    /**< input samples buffer */
```
this is fine for now, but its not future proof for future formats for sure

fuyuntsuo replied:
this is library's audio format configuration.

cujomalainey commented on tools/topology/platform/intel/intel-generic-dmic-kwd.m4:
```diff
@@ -42,7 +42,7 @@ ifdef(`DMIC_DAI_LINK_48k_NAME',`',define(DMIC_DAI_LINK_48k_NAME, `dmic01'))
 
 # DMICPROC is set by makefile, available type: passthrough/eq-iir-volume
-ifdef(`DMICPROC', , `define(DMICPROC, passthrough)')
+ifdef(`DMICPROC', , `define(DMICPROC, nr)')
```
please dont modify existing pipelines, either add switches for your config or fork and make your own

fuyuntsuo replied:
which one would be preferable? a switch or a new sof-tgl-max98373-rt5682-igonr.m4?

cujomalainey replied:
i think copying is better in this case for the pipeline level m4, and add switches to top level m4

fuyuntsuo replied:
Thanks! This looks better in this way.

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,523 @@
+		y = audio_stream_write_frag_s32(sink, idx_out);
+		*y = (cd->out[cd->out_rpt]);
+
+		if (cd->config.igo_params.dump_data == 1 && out_nch > 1) {
```
we might need to dump audio in/out buffer on the bad DUT. igo_params can be updated via ipc.

cujomalainey replied:
Please hide behind DEBUG config flag then, we dont want userspace accidentally triggering this

lgirdwood replied:
and you are also checking this in your loop. Loops should always be optimal.

fuyuntsuo replied:
@cujomalainey okay. Now the codes are hidden by CONFIG_DEBUG flag.
@lgirdwood thanks! moved the checking outside of loop.

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,523 @@
+	}
+}
+
+static inline int32_t set_func(struct comp_dev *dev)
```
okay. will change to `set_capture_func` because it's used to capture audio data.

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,523 @@
+		 sizeof(struct sof_ipc_comp_process), ipc_igo_nr,
+		 sizeof(struct sof_ipc_comp_process));
+
+	dev->state = COMP_STATE_INIT;
```
thanks. it's removed.

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,523 @@
+	ret = comp_init_data_blob(cd->model_handler, bs, ipc_igo_nr->data);
+	if (ret < 0) {
+		comp_cl_err(&comp_igo_nr, "igo_nr_new(): comp_init_data_blob() failed.");
+		rfree(dev);
```
okay.

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,523 @@
+
+static const struct comp_driver comp_igo_nr;
+
+uint8_t igo_heap[IGO_MEM_SIZE];
```
We've tried to allocate 40KB on heap but failed. So we put this working buffer to bss.

Besides, this component is not expected to be instantiated more than once because it takes 250MCPS but TGL get only 400MHz.

cujomalainey replied:
Hmm @lgirdwood thoughts? If this is needed at least then we should add a mechanism to prevent second instantiation and error out

lgirdwood replied:
@fuyuntsuo how did you try an allocate ? This does make it a singleton and prevents us optimising the physical memory usage (i.e to a certain bank or memory type). This bss is also non static and non aligned (for cache).

fuyuntsuo replied:
I tried the following two approaches:

1. rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, IGO_MEM_SIZE) ==> DSP failed to add widget, reply 0
2. rzalloc(SOF_MEM_ZONE_BUFFER, 0, SOF_MEM_CAPS_RAM, IGO_MEM_SIZE); ==> DSP panic, out of memory

From memory.h it seems that HEAP_BUFFER_SIZE has 0x80000 = 524,288 Bytes in total. But I still can't allocate 40KB in this region.

lgirdwood replied:
@fuyuntsuo can you try rballoc()
```c
/**
 * Similar to rmalloc(), guarantees that returned block is zeroed.
 *
 * @note Do not use  for buffers (SOF_MEM_ZONE_BUFFER zone).
 *       rballoc(), rballoc_align() to allocate memory for buffers.
 */
void *rzalloc(enum mem_zone zone, uint32_t flags, uint32_t caps, size_t bytes);

/**
 * Allocates memory block from SOF_MEM_ZONE_BUFFER.
 * @param flags Flags, see SOF_MEM_FLAG_...
 * @param caps Capabilities, see SOF_MEM_CAPS_...
 * @param bytes Size in bytes.
 * @param alignment Alignment in bytes.
 * @return Pointer to the allocated memory or NULL if failed.
 */
void *rballoc_align(uint32_t flags, uint32_t caps, size_t bytes,
		    uint32_t alignment);

/**
 * Similar to rballoc_align(), returns buffer aligned to PLATFORM_DCACHE_ALIGN.
 */
static inline void *rballoc(uint32_t flags, uint32_t caps, size_t bytes)
{
	return rballoc_align(flags, caps, bytes, PLATFORM_DCACHE_ALIGN);
}
```

paulstelian97 replied:
+1, `rballoc` in general permits allocating larger blocks of memory, which don't _have_ to be buffers. `rmalloc`/`rzalloc` allocates small blocks of memory which are useful for various simple data structures.

fuyuntsuo replied:
thanks! @lgirdwood @paulstelian97 @cujomalainey it works with `rballoc`.

fuyuntsuo replied:
Here is another issue to discuss: 

In addition to the working buffer `igo_heap`, in static library we still have around 6xx KB global data storing NN coefficient. Typically we keep  these coefficient in external storage, like flash, and then DMA to RAM when the component initializes.

However there is no such external storage on TGL, so we have no choice but keep them in global data region. Do you have any suggestion?

paulstelian97 replied:
Constant data should be fine in the data/rodata (put a const in front of the array). Just ensure that it's only there when you have the kconfig enabled.

fuyuntsuo replied:
That's for sure. The subdirectory won't be included when the feature is disabled in Kconfig.

lgirdwood replied:
@fuyuntsuo there is existing tooling and example code for coefficient loading at runtime. Look at SRC, IIR, FIR, EQ etc for examples. @singalsu is best person to ask questions here on loading coefficients.

fuyuntsuo replied:
@lgirdwood if you are referring to the run-time tuning tool, a.k.a. sof-ctl, it's indeed helpful for coefficient at the level of tens or hundreds of bytes. However the coefficient in this component is more than 600KB. If applying the same mechanism, it will obsoletely impose heavy burden to IPC during component prepare; moreover, it might starve IPC from other components. I am not sure whether this is a good idea.

lgirdwood replied:
@fuyuntsuo got you - 600kB is quite large for IPC loading. So two options (depends if runtime tuning from host is needed).

1) Build in the coefficient at build time (ok if we dont need to tune and modify them from host at runtime).
2) create an extra source pipeline and buffer into your component and connect it with topology. That way an ALSA compressed PCM could be used to send coefficient data at runtime from tuning tools.

singalsu replied:
IPC is pretty fast, have you tried how long pass of 600 kB via binary control takes? I haven't had problems to send some kilobytes of data for e.g. beamformer. The speed would also increase due to less overhead if the IPC size 384 bytes is increased.

lgirdwood replied:
@singalsu please try an update this, IIRC there is no hard limit here why we can take up more of the mailbox.

fuyuntsuo replied:
I agree. Depend on the physical clock rate of IPC, transmitting hundreds of KB upon component prepare might be feasible. 

But this is a bit out of the scope of this PR. If 600KB IPC passing is supported, we can treat the host driver as external storage and keep anything (text, rodata, etc.) there until SoF firmware needs them. It will be a huge change of framework imo. Currently we just need the coefficient to be linked statically according the Kconfig setting :)

fuyuntsuo replied:
@lgirdwood we would go for 1 because NN coefficients are fixed; run-time tuning is not expected during inference.

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,523 @@
+	}
+}
+
+static void igo_nr_bypass_s32(struct igo_nr_comp_data *cd,
```
these function have been renamed to `igo_nr_capture_s32`. they are doing more than pure copy function.

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,523 @@
+		cd->config = *p_config;
+
+		comp_info(dev, "config changed");
+		comp_info(dev, "  igo_params_ver       %d",
```
okay. done.

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,523 @@
+
+    /* Run algorithm if buffers are available */
+	if ((cd->in_wpt == IGO_FRAME_SIZE) && (cd->out_rpt == IGO_FRAME_SIZE) {
+		comp_dbg(dev, "algo run");
```
this debug code has been removed.

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,523 @@
+		       sink_frames * cl.sink_frame_bytes);
+
+    /* Run algorithm if buffers are available */
+	if ((cd->in_wpt == IGO_FRAME_SIZE) && (cd->out_rpt == IGO_FRAME_SIZE) {
```
In average the component should see 256 frames per function call, as long as the topology is configured correctly. So running twice is not possible.

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,523 @@
+
+		if (cd->config.igo_params.nr_bypass) {
+			memcpy_s(cd->out, IGO_FRAME_SIZE * sizeof(int16_t),
+				 cd->in, IGO_FRAME_SIZE * sizeof(int16_t));
```
Yes. `IgoLibProcess` takes only 16-bit audio data so these two working buffer are fixed to a particular format.

paulstelian97 replied:
We must ensure via the params() method that only 16-bit formats are allowed then.

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,523 @@
+				 cd->in, IGO_FRAME_SIZE * sizeof(int16_t));
+		} else {
+			cd->igo_stream_data_in.data = cd->in;
+			cd->igo_stream_data_in.data_width = IGO_DATA_16BIT;
```
These codes are used to configure library's audio data format, not SOF's audio data format. 

For 32-bit audio passing to the component, it will be truncated to 16-bit inside `igo_nr_capture_s32` before sending to `IgoLibProcess`.

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,523 @@
+			cd->igo_stream_data_in.data = cd->in;
+			cd->igo_stream_data_in.data_width = IGO_DATA_16BIT;
+			cd->igo_stream_data_in.sample_num = IGO_FRAME_SIZE;
+			cd->igo_stream_data_in.sampling_rate = 16000;
```
These are used to configure library's audio format. If the real data_width and sample_num doesn't fit in, the wrapper codes will handle that.

cujomalainey replied:
please show me where the wrapper would reject, i don't seen anything in the params or prepare callbacks that would reject s24/s32/float or other channel counts or rates

fuyuntsuo replied:
I rewrite the code and now the wrapper would reject anything other than sample rate = 16000 and s16/s32 in `set_capture_func`.

iuliana-prodan replied:
In `set_capture_func` you only filter the frame format (s16 or s32). Where is the sample rate filtered?

fuyuntsuo replied:
I bet Github hasn't updated yet when you see the code. Ping you in another issue.

cujomalainey replied:
Github should update <5s after you force push the branch related to the PR

fuyuntsuo replied:
I am not familiar with Github backend. But out of curious, is it necessary to re-request review every time I update the PR? Usually I reply comments, rewrite code, force push, and re-request review at last. If it's not standard flow please let me know.

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,523 @@
+			cd->igo_stream_data_out.data = cd->out;
+			cd->igo_stream_data_out.data_width = IGO_DATA_16BIT;
+			cd->igo_stream_data_out.sample_num = IGO_FRAME_SIZE;
+			cd->igo_stream_data_out.sampling_rate = 16000;
```
Please refer to the reply above.

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,523 @@
+	struct comp_buffer *sinkb;
+	int32_t ret;
+
+	comp_info(dev, "igo_nr_prepare()");
```
thanks. done.

dbaluta commented on src/audio/Kconfig:
```diff
@@ -338,6 +338,15 @@ config COMP_CODEC_ADAPTER
+	help
+	  This option enables intelliGo non-speech noise reduction. The propitiatory static library
+	  libigonr.a is needed. The algorithm runs in 16kHz domain with 256 samples in a frame.
+	  GCC is currently not supported because in the library lots of Xtensa intrinsics have been
```
typo: proprietary.

dbaluta replied:
Does it work on HIFI4? :)

fuyuntsuo replied:
thanks. the typo has been fixed.

Currently we only support HiFi3, more specifically `HiFi3 cavs`. The static library is built on that toolchain so SOF image must be built with the same toolchain too.

lgirdwood replied:
See my comment above on kconfig help text. This needs your company link and contact page so users can aks questions or request new architectures etc.

fuyuntsuo replied:
Sure. Thanks for providing a helpful suggestion!

cujomalainey commented on src/audio/Kconfig:
```diff
@@ -338,6 +338,16 @@ config COMP_CODEC_ADAPTER
+config COMP_IGO_NR
+	depends on XCHAL_HAVE_HIFI3
+	bool "IGO NR component"
+	default n
```
add depends on `FORMAT_S16LE` since you are locked to a particular format

fuyuntsuo replied:
Actually we accept S16_LE and S32_LE in the wrapper. In the case of S32_LE we just right shift the data by 16 bits.

I will rewrite the code to make it more clear.

cujomalainey commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,521 @@
+	ret = IgoLibGetInfo(&cd->igo_lib_info);
+	if (ret != IGO_RET_OK) {
+		comp_cl_err(&comp_igo_nr, "igo_nr_new(): IgoLibGetInfo() Failed.");
+		rfree(dev);
```
inittialize cd to NULL and you can use the same goto here and above since free(NULL) -> no-op

fuyuntsuo replied:
Done. Thanks for the hint!

lgirdwood commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,521 @@
+
+		cd->in_wpt++;
+		idx_in += in_nch;
+	}
```
memcpy ? IIRC I think there is a buffer API that does this and also takes in the wrap.

fuyuntsuo replied:
this is more like deinterleaving data one by one. not pure memcpy and no wrap around.

cujomalainey replied:
I feel like we should make this component expect 1 channel in, using something like the selector, at an absolute minimum we need a comment

fuyuntsuo replied:
Even for more than 1 channel in, this component still selects the first channel to process.

cujomalainey replied:
and just copy the rest? that seems confusing to users, i feel like we should only consume what we intend to process

fuyuntsuo replied:
@cujomalainey my understanding is that even if the component process the first channel and bypass the other channels, it still has to consume all the channels. Consume/produce function are expected to take care of rpt/wpt update for audio buffer. It's impossible to consume the first channel only because data are interleaved.

singalsu replied:
You can read source buffer with channels count stride and just update to consumed with the size that includes the omitted channel(s).

fuyuntsuo replied:
@singalsu sorry I am a bit confusing. the code is doing exactly what you are saying. what is the difference?

singalsu replied:
Sorry, I missed that while just looked the previous comments. Please ignore my comment!

fuyuntsuo replied:
No problem!

lgirdwood replied:
@fuyuntsuo it's probably worth a comment at the start of each of these function explaining the deinterleaving and why it's needed. That way everyone should be aligned.

fuyuntsuo replied:
@lgirdwood agree. will put comments to ensure everyone is aligned.

lgirdwood commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,521 @@
+
+		cd->out_rpt++;
+		idx_out += out_nch;
+	}
```
ditto, the above comments apply here too.

fuyuntsuo replied:
any keyword? I can't locate the any code doing deinterleave.

lgirdwood replied:
oh, it's maybe that we only have non interleave copy. That's probably a todo.

lgirdwood commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,521 @@
+
+	comp_info(dev, "igo_nr_get_igo_params()");
+
+	struct sof_igo_nr_config *p_config = comp_get_data_blob(cd->model_handler, NULL, NULL);
```
p_config should be declared before the comp_info()

fuyuntsuo replied:
fixed. thanks!

fuyuntsuo commented on src/audio/Kconfig:
```diff
@@ -338,6 +338,16 @@ config COMP_CODEC_ADAPTER
 	  different codecs and use them in parallel.
 
+config COMP_IGO_NR
+	depends on XCHAL_HAVE_HIFI3
```
Indeed there is no such Kconfig value. I removed the dependency since it should not depends on specific toolchain.

cujomalainey commented on src/audio/igo_nr/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+
+add_local_sources(sof igo_nr.c)
+
+cmake_policy(SET CMP0079 NEW)
```
why is this needed?

fuyuntsuo replied:
These code, which were used to get rid of error during linking to the static library, will be removed. Because I just notice that SoF framework provides `sof_add_static_library` which is much more elegant.

cujomalainey reviewed (changes requested):
getting better, but you are still missing your checks in either `prepare` or `params` to verify the current hardware params are valid for your capabilities.

fuyuntsuo commented:
> getting better, but you are still missing your checks in either `prepare` or `params` to verify the current hardware params are valid for your capabilities.

Inside `set_capture_func` of `igo_nr_prepare`, we will check the sample rate and frame format.

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,534 @@
+
+	cd = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (!cd)
+		goto err;
```
cd is initialized to NULL so rfree(NULL) is no-op. then we can reuse the same goto. (appreciate hint from @cujomalainey)

iuliana-prodan replied:
@fuyuntsuo you're right, it should be ok.

iuliana-prodan commented:
> > getting better, but you are still missing your checks in either `prepare` or `params` to verify the current hardware params are valid for your capabilities.
> 
> Inside `set_capture_func` of `igo_nr_prepare`, we will check the sample rate and frame format.

Maybe I'm missing something, but I don't see where do you check for sample rate?
In `set_capture_func` you filter the frame format.
In `igo_nr_lib_init`, called from `igo_nr_prepare`, you just set the number of channels to 1 and the sample rate to 16000. 
I don't see any validation.

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,543 @@
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer,
+				  sink_list);
+
+	switch (sourceb->stream.rate) {
```
@iuliana-prodan This is the place where sample rate being filtered.

cujomalainey replied:
hmm, this seems confusing from a user perspective. To disable/enable NC based on HW params. Typically what we would do is have a amixer switch, NC on/off. That way NC disable/enable does not require pipeline restart. That way your module can also have a consistent set of params that it accepts. @chiang831 what are your thoughts?

cujomalainey replied:
also @lgirdwood would like your thoughts here as well

fuyuntsuo replied:
@cujomalainey do you mean that the component should not reject the audio stream, whose parameters are not supported by the component yet, but instead bypass them from input to output?

cujomalainey replied:
I'm saying we should have a amixer switch to control whether algorithm is on. If switch is on, all hw_params not supported should be rejected with EINVAL. If switch is off then we use pass through channels but we could allow more hw_params through based on our sorted functions. I am just saying the behaviour of the pipeline should not be based of the params, but on the controls as those explain better to userspace the expected experience.

lgirdwood replied:
Yeah, lets not make this magically switch on/off via an indirect sample rate switch. It has to be obvious to users if it's on/off so having a kcontrol or rejecting non 16kHz rates with an error would be required.

fuyuntsuo replied:
thanks @cujomalainey @lgirdwood. just a few questions before jumping to code rewrite. 

1. Sample rate is a HW parameter decided by Codec, but bit width (S32_LE etc.) is not, rite?
2. To reject unsupported parameter when amixer on, or to pass through when amixer off, requires that the component keep track of amixer on/off status. Is there an existing flag? I couldn't find any reference inside tdfb or src component.

cujomalainey replied:
format (i.e. s32_le) is a parameter that you have to check.

you will get a callback on `.cmd` and follow the examples to parse our whether it is a get/set. You will store the state in your comp_data struct and use that to track whether you are activated.

fuyuntsuo replied:
Should sampling rate (i.e., `arecord -r16000`) be also checked too? that's my original thought.

Seems like there is no existing amixer switch status inside the component. How about we check `dev->state`  in prior? That is, if it's `COMP_STATE_ACTIVE` then it returns EINVAL when sample rate != 16000, if it's not `COMP_STATE_ACTIVE` the it just let everything to pass through.

lgirdwood replied:
The hw_params will set the rate and format and these will all be known at trigger(start) so can be easily rejected as not supported if the rate and format is incorrect (I would also print something in the log saying the rate or format is not valid).
Additionally the kcontrol could then switch ON/OFF the processing (after the params have been validated above).

cujomalainey replied:
@fuyuntsuo your hw_params are passed during the `params` callback. If you choose not to do it during that callback then you should do it in `prepare` which comes after `params.` You don't need to check your state, simply checking on prepare will make sure you are in the correct configuration.

Additionally to make things easier, we can say a switch change will not take effect until the pipeline is restarted (i.e. copy does not handle any ipc config change)

fuyuntsuo replied:
@cujomalainey okay. Here is the change I will start to implement, please help review the flow. 

1. Add a NC control as a volsw kcontrol. Keep it as an ON/OFF switch inside igo_nr component. (default=OFF)
2. Validate hw_params in `.params` callback. If switch ON, reject with `EINVAL`. If switch OFF, pass through all channels.
3. Move other algorithm parameters to UCM TLV binary, which will be passed to library via `SOF_CTRL_CMD_BINARY`.

In this way we should be able to cover all the tuning paths (i.e., sof-ctl, UCM, kcontrol) for igo_nr.

lgirdwood replied:
Re 2) I would reject all invalid configs regardless if switch was ON/OFF unless @cujomalainey we have use case for the pipeline with processing being off ?

cujomalainey replied:
@fuyuntsuo please use the explicit switch control, we don't need volume slider. 

@lgirdwood the only case is trying to record in 48k with processing off. But I am wondering if maybe we should use mux due to different rates.

e.g.
                  IGO <- DAI(16k)
                 /
PCM <- mux
                 \
                  DAI(48k)

cujomalainey replied:
Alternative is put SRC in front of IGO on 48k DAI

lgirdwood replied:
@cujomalainey thanks got you, Btw we have dynamic pipeline now so we could just unload the IGO pipe and load a passthrough pipe for 48kHz capture from same DAI endpoint ? Both pipes would have a different PCM device and would be mutually exclusive ?

lgirdwood replied:
@fuyuntsuo I think this now looks correct for IGO.
@cujomalainey are you fine with using dynamic pipelines to switch between IGO and a volume passthrough pipe from the same endpoint ?

cujomalainey replied:
Hmm, that might be a bit annoying with the way we setup our config but we could get around it, we are currently expecting to be able to toggle with a UCM modifier and restarting the same pipeline. But since we can guarantee mutual exclusivity we can save memory? This sounds very beneficial. Lets sync out of band.

fuyuntsuo replied:
@lgirdwood @cujomalainey sorry for late reply. I am still trying to figuring out a way to load the mixer kcontrol to igo_nr component. All the other reference topologies seems to be `pga` type component but ours is `feature` type.  Following is the observation with current code and topology:

1. Dmesg log [here](https://github.com/thesofproject/sof/files/6381318/dmesg.txt) - error: no scontrol for widget IGONR4.0 / error: process loading failed.
2. By removing mixer part (L63-65 of igo_nr.m4), we can capture the kcontrol by `amixer -c0 | grep -i igo` on TGL DUT. But the value change doesn't propagate to SOF FW.
3. By changing igo_nr from `feature` type to `pga` type, the DUT hangs.

@marc-hb @lbetlej @mmaka1 @plbossart @ranj063 @dbaluta @paulstelian97 @iuliana-prodan @singalsu 
Please advise. I am wondering if these mixer kcontrol thing only fits `pga` type component... Thanks in advnace!

cujomalainey commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,543 @@
+	sourceb = list_first_item(&dev->bsource_list,
+				  struct comp_buffer, sink_list);
+	sinkb = list_first_item(&dev->bsink_list,
+				struct comp_buffer, source_list);
```
these two variables are set and never used

fuyuntsuo replied:
thanks. just removed.

singalsu commented on tools/topology/platform/intel/intel-generic-dmic-kwd.m4:
```diff
@@ -42,7 +42,20 @@ ifdef(`DMIC_DAI_LINK_48k_NAME',`',define(DMIC_DAI_LINK_48k_NAME, `dmic01'))
 # DMICPROC is set by makefile, available type: passthrough/eq-iir-volume
-ifdef(`DMICPROC', , `define(DMICPROC, passthrough)')
+ifdef(`IGO',
+`define(DMICPROC, nr)',
```
should there be a pipe-nr-capture.m4 in the commit, I wonder if it's missing or I just didn't find it. Would the pipeline macro name pipe-igonr-capture.m4 be better?

fuyuntsuo replied:
This PR consists of two commits: 
`audio: igo_nr: Added intelliGo noise reduction wrapper code and topology` is the generic codes and topology.
`topology: Added IGO_NR in sof-tgl-max98373-rt5682 topology.` is the customized topology to add igo_nr into that topology

`pipe-nr-capture.m4` is inside the first commit.

fuyuntsuo replied:
agree. I will change the name to `pipe-igonr-capture.m4`.

singalsu commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,537 @@
+
+	for (i = 0; i < src_frames; i++) {
+		x = audio_stream_read_frag_s32(source, idx_in);
+		cd->in[cd->in_wpt] = (*x) >> 16;
```
Perhaps use round here?

fuyuntsuo replied:
we always take the MSB 16bit regardless of input frame format being S32_LE, S24_LE, or S16_LE.

singalsu replied:
The shift without round when going from larger world length to smaller creates a tiny DC offset to signal but it's up to you to decide.

fuyuntsuo replied:
Ah thanks for the reminding. We are aware of quantization noise induced by bit shift. It's totally fine for our algorithm.

fuyuntsuo replied:
@singalsu my apologies. the tiny DC offset in your statement is due to biased rounding, not quant noise. We will have a short discussion about it internally.

fuyuntsuo replied:
@singalsu we decide to add a unbias shift prior to copying to data buffer. thanks!

singalsu replied:
There's e.g. macro Q_SHIFT_RND() in sof/audio/format.h that does it.

fuyuntsuo replied:
thanks. replaced with Q_SHIFT_RND().

singalsu commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,537 @@
+		comp_info(dev, "set_capture_func(), SOF_IPC_FRAME_S32_LE");
+		cd->igo_nr_func = igo_nr_capture_s32;
+		break;
+	default:
```
Do you have plan to support S24_4LE? It's quite straightforward with shifts like done with S32_LE.

fuyuntsuo replied:
sure we can. will update then.

singalsu commented on tools/topology/m4/igo_nr.m4:
```diff
@@ -0,0 +1,66 @@
+divert(-1)
```
This should go to topology commit.

fuyuntsuo replied:
actually both commits have topology files: one is general change and the other is customized change for sof-tgl-max98373-rt5682.m4. `igo_nr.m4` has to be there for arbitrary topology so I put it inside the audio commit.

singalsu commented on tools/topology/sof/pipe-nr-capture.m4:
```diff
@@ -0,0 +1,89 @@
+# Capture Passthrough Pipeline and PCM
```
This also to topology commit (I wondered where it is). The name could be perhaps prefixed with "igo" since this is not a generic NR solution, needs the IGO library and license.

fuyuntsuo replied:
agree. will change the file name to `pipe-igonr-capture.m4` then.

cujomalainey commented:
please also address checkpatch comments, some of which it looks like are irrelevant, but most are not

fuyuntsuo commented:
> please also address checkpatch comments, some of which it looks like are irrelevant, but most are not

Okay. I've cleaned up the error/warning that seems correctable. Please let me know if any of remaining ones matters.

lgirdwood commented on tools/topology/sof-smart-amplifier.m4:
```diff
@@ -122,7 +125,7 @@ dnl     time_domain, sched_comp)
 PIPELINE_PCM_ADD(sof/pipe-smart-amplifier-playback.m4,
 	SMART_PB_PPL_ID, SMART_PCM_ID, SMART_PB_CH_NUM, s32le,
-	1000, 0, SMART_AMP_CORE,
+	SMART_AMP_PERIOD, 0, SMART_AMP_CORE,
```
@fuyuntsuo sorry, not following the changes to SMART_AMP here and how it relates ?

fuyuntsuo replied:
The long process time (aprox. 8ms) and frame size (16ms in 16k sample rate) of igo_nr starve the scheduler and results in SMART_AMP underflow, ending up with smart_amp component reset and close. So we have to increase the buffer size of SMART_AMP.

lgirdwood replied:
Got you, probably worth mentioning this in the comments.

fuyuntsuo replied:
sure. will put more description next to SMART_AMP_PERIOD define.

lgirdwood commented on src/include/user/igo_nr.h:
```diff
@@ -0,0 +1,63 @@
+	uint32_t nr_mode1_pp_param10;
+	uint32_t nr_mode3_floor;
+	uint32_t nr_mode1_pp_param53;
+};
```
This should be packed if it's visible to userspace as it guarantees the ABI will align no matter the architecture.

fuyuntsuo replied:
These parameters are visible to userspace. For example, the sof-ctl needs this structure to control igo_nr component.

fuyuntsuo replied:
btw, packed was removed because of checkpatch warning.

lgirdwood replied:
please ignore checkpatch warning here (we need to fix), the older XCC compiler does not understand the new syntax for ```packed``` and tends to produce errors. We need to use the old packed syntax (age just ignore the checkpatch errors).

fuyuntsuo replied:
@lgirdwood  understand. thanks for explanation!

lgirdwood commented on src/include/user/igo_nr.h:
```diff
@@ -0,0 +1,63 @@
+	struct IGO_PARAMS igo_params;
+
+	int16_t data[];
+};
```
ditto

fuyuntsuo replied:
please refer to my comment above.

lgirdwood reviewed (commented):
I think from my POV I just need a little more in terms of comments describing the data processing flows and data formats and conversions.

cujomalainey commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,602 @@
+// SPDX-License-Identifier: BSD-3-Clause
```
check patch is still complaining about these comments

fuyuntsuo replied:
checkpatch also complain even if it is changed to `/* */`.  

`WARNING: Improper SPDX comment style for 'src/audio/igo_nr/igo_nr.c', please use '//' instead`
`#11: FILE: src/audio/igo_nr/igo_nr.c:1:`
`+/*SPDX-License-Identifier: BSD-3-Clause`

`WARNING: Missing or malformed SPDX-License-Identifier tag in line 1`
`#11: FILE: src/audio/igo_nr/igo_nr.c:1:`
`+/*SPDX-License-Identifier: BSD-3-Clause`

so I take other components as reference and remain `//` .

cujomalainey replied:
You need a space between `/*` and your first character. That is why it is upset in your second case.

fuyuntsuo replied:
I get same warning even with that space added. Let me try to force push here to see if the result is same or not.

cujomalainey replied:
turns out checkpatch and post commit script don't agree with each other. Use `//` sorry for the noise. @lgirdwood should we fix the post commit hook?

fuyuntsuo replied:
@cujomalainey it's okay. guess that's why everyone is using `//`.

lgirdwood replied:
yep, checkpatch is very Linux specific and should be ignored here. It does need some TLC as it also brings up requests that break XCC.  @marc-hb fyi - when time permits.

marc-hb replied:
I recently added `--ignore C99_COMMENT_TOLERANCE` in commit eb459078f302 because I got tired of bumping into this unwritten SOF policy and wasting precious code review time. As you've noticed, this is in contradiction with another checkpatch warning. Checkpatch is not perfect and is expected to have some contradictions.

https://www.kernel.org/doc/html/v4.12/process/5.Posting.html#sending-the-patch
> Are you sure your patch is free of silly mistakes? You should always run patches through scripts/checkpatch.pl and address the complaints it comes up with. Please bear in mind that **checkpatch.pl**, while being the embodiment of a fair amount of thought about what kernel patches should look like, **is not smarter than you.** **If fixing a checkpatch.pl complaint would make the code worse, don’t do it.**


Here's another, much more painful contradiction:

```
src/test.c:6: WARNING: break quoted strings at a space character
+	f("very long line very long line very long line very long line"
+	  "very long line very long line very long line very long line");
```
but:
```
src/test.c:5: WARNING: line length of 133 exceeds 100 columns
+	f("very long line very long line very long line very long line very long line very long line very long line very long line");
```
... and `--strict` is not even turned on for this one. Any way out of it?

As long as these contradictions are _rare_ then everything fine. We don't add new files every day whereas any line of code could have a C99 comment.

If something becomes too frequent then we go and fix it: https://github.com/thesofproject/sof/commits/main/scripts/checkpatch.pl

lgirdwood reviewed (commented):
Almost ready to merge. Just a few items to close.

lgirdwood reviewed (approved):
LGTM, but lets align on the pipeline switching between IGO and 48kHz just in case we have to change anything here.

cujomalainey reviewed (dismissed):
params check looks good to me, thanks

cujomalainey commented:
revoking approval as there is a topology bug outstanding

lgirdwood commented:
@cujomalainey @paulstelian97 can someone highlight the topology error so @fuyuntsuo can fix it :)

cujomalainey commented:
@fuyuntsuo please post error as requested so we can try and help you resolve it

fuyuntsuo commented:
> @fuyuntsuo please post error as requested so we can try and help you resolve it

Description: The `pipeline-igonr-capture.m4` and `igo_nr.m4` failed to add kcontrol in IGO_NR widget. The dmesg log is [here](https://github.com/thesofproject/sof/files/6388595/dmesg.txt).

Detail: It seems that only `pga` type widget can add simple kcontrol (op."ctl" 259). The `effect` type widget, like igo_nr, cannot add such kcontrol (as in L63:igo_nr.m4). Please help point a possible way out.

juimonen commented:
@fuyuntsuo oh currently you can't add 2 different type controls (bytes and mixer in your case) to a single widget. This is an alsa peculiarity that you can add only several _same_ type controls. I have a PR/patch for this: https://github.com/thesofproject/linux/pull/2423, that hasn't gained too much traction -> I should push it more...

lgirdwood commented:
> @fuyuntsuo oh currently you can't add 2 different type controls (bytes and mixer in your case) to a single widget. This is an alsa peculiarity that you can add only several _same_ type controls. I have a PR/patch for this: [thesofproject/linux#2423](https://github.com/thesofproject/linux/pull/2423), that hasn't gained too much traction -> I should push it more...

@juimonen Please keep pushing this - we have a use case now for it. @plbossart fyi.
@ranj063  we also need dynamic pipelines merged for IGO support (to support switching between IGO and passthrough capture using the same DMIC).
@fuyuntsuo is it possible to keep IGO always enabled (by default) for the time being and comment out the ON/OFF kcontrol. We can re-enable this one the kernel PR is merged.

@cujomalainey does this sound ok atm ? We would have a dependency on 2 kernel patches.

fuyuntsuo commented:
> @fuyuntsuo oh currently you can't add 2 different type controls (bytes and mixer in your case) to a single widget. This is an alsa peculiarity that you can add only several _same_ type controls. I have a PR/patch for this: [thesofproject/linux#2423](https://github.com/thesofproject/linux/pull/2423), that hasn't gained too much traction -> I should push it more...

Thanks! @juimonen This timely information is very important. Guess that's' why I've tried ENUM and MIXER but both fail too.

fuyuntsuo commented:
> > @fuyuntsuo oh currently you can't add 2 different type controls (bytes and mixer in your case) to a single widget. This is an alsa peculiarity that you can add only several _same_ type controls. I have a PR/patch for this: [thesofproject/linux#2423](https://github.com/thesofproject/linux/pull/2423), that hasn't gained too much traction -> I should push it more...
> 
> @juimonen Please keep pushing this - we have a use case now for it. @plbossart fyi.
> @ranj063 we also need dynamic pipelines merged for IGO support (to support switching between IGO and passthrough capture using the same DMIC).
> @fuyuntsuo is it possible to keep IGO always enabled (by default) for the time being and comment out the ON/OFF kcontrol. We can re-enable this one the kernel PR is merged.
> 
> @cujomalainey does this sound ok atm ? We would have a dependency on 2 kernel patches.

Absolutely. I've force-pushed the code with the following changes:
1. Added NR on/off switch handler via `SOF_CTRL_CMD_ENUM` and `SOF_CTRL_CMD_SWITCH`.
2. Sample rate check moved from `.prepare` to `.params` (L365:igo_nr.c)
3. Set NR on by default (L748:igo_nr.c)
4. Removed simple mixer kcontrol from igo_nr widget.

@cujomalainey Please review again and see if there is any issue not resolved yet.

lgirdwood commented:
> @ranj063 we also need dynamic pipelines merged for IGO support (to support switching between IGO and passthrough capture using the same DMIC).

@cujomalainey Just confirmed this was merged recently but would need backporting to TGL kernel.

cujomalainey commented:
> > @ranj063 we also need dynamic pipelines merged for IGO support (to support switching between IGO and passthrough capture using the same DMIC).
> 
> @cujomalainey Just confirmed this was merged recently but would need backporting to TGL kernel.

Understood, will file a bug internally to get that integration tracked. It might be worth just reving a release and picking up some of the other modules integrated.

cujomalainey commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+	int32_t ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_ENUM:
```
just curious why you have both enum and switch? is one broken?

fuyuntsuo replied:
Nope. I just added both of them during the topology debugging. No harm to keep them there in case we might need one of them in the future.

lgirdwood replied:
Lets remove the enum since ALSA will map the switch as an ON/OFF type control. @fuyuntsuo this can be done as an update.

fuyuntsuo replied:
This issue has been resolved in #4152.
@lgirdwood I don't quite follow. SWITCH is the ON/OFF type control but ENUM is the multiple choice type control. What's the reason to keep SWITCH only? Will ENUM be deprecated in the future?

cujomalainey commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+	cd->sink_frames_max = cd->sink_frames + 10;
+	cd->frames = MAX(cd->source_frames_max, cd->sink_frames_max);
+
+	comp_info(dev, "igo_nr_params(), source_rate=%u, sink_rate=%u, source_frames_max=%d, sink_frames_max=%d",
```
reduce this and the line below to dbg please, otherwise they will be quite noisy

fuyuntsuo replied:
fixed. thanks!

cujomalainey commented:
I just nee to understand the reason for both enum and switch, once we clear that I think we are good to go

fuyuntsuo commented:
> I just nee to understand the reason for both enum and switch, once we clear that I think we are good to go

just think no harm to keep both get/set value handlers.

cujomalainey reviewed (approved):
component looks good, Intel please verify you are ok with topology changes. Thanks for keeping with this.

fuyuntsuo commented:
> component looks good, Intel please verify you are ok with topology changes. Thanks for keeping with this.

@cujomalainey thanks! Just updated the copyright.

cujomalainey commented:
I would ideally like to get one more review, @ranj063 mind doing a review?

fuyuntsuo commented:
Seems like the reviewer has not responded yet. Is there anything I can do?

I just check the CI messages again and fix most of the complains. (The remaining complains have been discussed before.)

Please let me know if there is any PR blocker other than this one.

lgirdwood commented:
@fuyuntsuo I will ping @ranj063 to take a look.
@kkarask @wszypelt @zrombel looks like a whole WHL series failure in CI, but this is unrelated. Good to go ?

ranj063 commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+	struct sof_ipc_comp_process *ipc_igo_nr =
+		(struct sof_ipc_comp_process *)comp;
+	struct comp_dev *dev;
+	struct comp_data *cd = NULL;
```
NULL assignemtn not necessary

fuyuntsuo replied:
With this NULL assignment, the memory allocations afterwards can share the same error handler.

That is,  when `if (!cd)` holds (L243:igo_nr.c), then it goes to the last and rfree(NULL) becomes nop.

ranj063 replied:
sorry not getting this point. Is (!cd) is true, that mean cd is NULL anyway

cujomalainey replied:
@fuyuntsuo none of the error paths before your alloc for `cd` jump to `err` so this NULL doesn't make a difference

lgirdwood replied:
Lets fix this incrementally.

fuyuntsuo replied:
Get it @ranj063 @cujomalainey. This issue has been resolved in #4152.

ranj063 commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+	if (ret < 0) {
+		comp_err(dev, "igo_nr_verify_params(): comp_verify_params() failed.");
+		return ret;
+	}
```
i'd remove the return ret here and just return ret outside the if

fuyuntsuo replied:
Agreed. Just moved outside.

ranj063 commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+		 cd->source_frames_max, cd->sink_frames_max);
+
+	/* The igo_nr supports sample rate 16000 only. */
+	switch (sourceb->stream.rate) {
```
do ou plan to support other formats in the future? if not, why not return -EINVAL is stream.rate != 16000 here?

fuyuntsuo replied:
Yes. We'd like to reserve the possibility of implementing SRC inside our library. However this is not prioritized now because SOF FW seems to support SRC component already.

ranj063 reviewed (commented):
> I would ideally like to get one more review, @ranj063 mind doing a review?

@cujomalainey working on it now.

ranj063 commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+				   int32_t max_size)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int32_t ret = 0;
```
init to 0 not necessary?

fuyuntsuo replied:
okay. just removed all of them.

ranj063 commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int32_t j;
+	int32_t ret = 0;
```
same here init not necessary

fuyuntsuo replied:
ok. thanks.

ranj063 commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+				   struct sof_ipc_ctrl_data *cdata)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int32_t ret = 0;
```
same here

fuyuntsuo replied:
okay. thanks.

ranj063 commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+			igo_nr_set_chan_passthrough(dev, ch);
+	}
+
+	return ret;
```
looks like we dont need ret at all. just remoe it and return 0?

fuyuntsuo replied:
the ret was missing. I add it back. thanks!

fuyuntsuo replied:
Yes you are right. Return 0 here is just fine. Thanks!

ranj063 commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+	int32_t ret = 0;
+
+	switch (cdata->cmd) {
+		break;
```
break?

fuyuntsuo replied:
sorry this is stupid. removed it.

ranj063 commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+		case IGO_NR_ONOFF_SWITCH:
+			igo_nr_set_chan(dev, cdata);
+			break;
+
```
remove unnecessary line

fuyuntsuo replied:
done.

ranj063 commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+		break;
+	}
+
+	return ret;
```
we should just return -EINVAL in the default cases, remove ret, and return 0 here.

fuyuntsuo replied:
the `igo_nr_set_chan` ret has been added back. so I keep it the same.

ranj063 commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+		comp_dbg(dev, "igo_nr_cmd_set_value(), SOF_CTRL_CMD_ENUM index=%d", cdata->index);
+		switch (cdata->index) {
+		case IGO_NR_ONOFF_SWITCH:
+			igo_nr_set_chan(dev, cdata);
```
why dont we check the return value of igo_nr_set_chan()?

fuyuntsuo replied:
thanks. I add it back.

ranj063 commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		ret = igo_nr_cmd_set_data(dev, cdata);
+		break;
```
return igo_nr_cmd_set_data(dev, cdata); here and all the cases below and remove ret

fuyuntsuo replied:
okay done.

ranj063 commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+	cd->igo_stream_data_out.sample_num = IGO_FRAME_SIZE;
+	cd->igo_stream_data_out.sampling_rate = 16000;
+
+	return 0;
```
looks like this function cannot error, do we need a return value?

cujomalainey replied:
good point, can lib init fail? we should check it if it can

fuyuntsuo replied:
okay. removed ret.

ranj063 commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+	cd->process_enable[0] = true;
+
+	ret = set_capture_func(dev);
+	return ret;
```
return set_capture_func(dev);

fuyuntsuo replied:
done.

ranj063 commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,804 @@
+
+static int32_t igo_nr_trigger(struct comp_dev *dev, int32_t cmd)
+{
+	int32_t ret = 0;
```
init not needed

fuyuntsuo replied:
done.

ranj063 commented on tools/topology/platform/intel/intel-generic-dmic-kwd.m4:
```diff
@@ -44,7 +44,7 @@ ifdef(`DMIC_DAI_LINK_16k_NAME',`',define(DMIC_DAI_LINK_16k_NAME, `dmic16k'))
 ifdef(`IGO',
 `define(DMICPROC, igonr)',
-ifdef(`DMICPROC', , `define(DMICPROC, passthrough)'))
+`ifdef(`DMICPROC', , `define(DMICPROC, passthrough)')')
```
should be in the previous patch?

fuyuntsuo replied:
sorry. wrong rebase. fixed.

ranj063 commented on tools/topology/platform/intel/intel-generic-dmic-kwd.m4:
```diff
@@ -81,6 +81,7 @@ PIPELINE_PCM_ADD(sof/pipe-`DMICPROC'-capture.m4,
 undefine(`PGA_NAME')
 undefine(`PIPELINE_FILTER1')
 undefine(`PIPELINE_FILTER2')
+undefine(`DMICPROC')
```
this one toom should be in the first patch

fuyuntsuo replied:
fixed. thanks.

ranj063 commented:
> @ranj063 we also need dynamic pipelines merged for IGO support (to support switching between IGO and passthrough capture using the same DMIC).

@lgirdwood the dynamic pipeline PR is merged to sof-dev. Switching just the pipelines for IGO to dynamic is also supprted in topology. But like we discussed, the dynamic pipelines patches are not upstream yet and also not part of the 5.4 kernel.

cujomalainey reviewed (commented):
@ranj063 thanks for catching that, I really need to get cppcheck into the CI for those kind of errors.

fuyuntsuo commented:
Despite the changes above, I also moved a struct (L32:igo_nr.c) and a macro (L30:igo_nr.c) from `igo_nr.h` to `igo_nr.c` because they are intended to be called inside `igo_nr.c` only.

ranj063 commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,811 @@
+static void igo_nr_get_igo_params(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
```
remove new line

fuyuntsuo replied:
done.

fuyuntsuo replied:
But I get this warning in post-commit check. Or you are saying the new line after comp_info?
![image](https://user-images.githubusercontent.com/80802098/117097762-21003880-ad9f-11eb-8e62-48e35a867b0c.png)

lgirdwood replied:
@fuyuntsuo this can be fixed incrementally as I'm going to tag v1.8-rc1 today. The newline needs to go after the variable declarations i.e. after line 601 and before 602.

fuyuntsuo replied:
Thanks! This issue has been resolved in #4152.

ranj063 commented:
> Despite the changes above, I also moved a struct (L32:igo_nr.c) and a macro (L30:igo_nr.c) from `igo_nr.h` to `igo_nr.c` because they are intended to be called inside `igo_nr.c` only.

Thanks for the update @fuyuntsuo. Just a few more minor comments

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,811 @@
+	err = igo_nr_verify_params(dev, pcm_params);
+	if (err < 0) {
+		comp_err(dev, "igo_nr_params(): pcm params verification failed.");
+		return -EINVAL;
```
Done. Thanks!

fuyuntsuo commented on src/audio/igo_nr/igo_nr.c:
```diff
@@ -0,0 +1,811 @@
+				  cdata->chanv[j].value);
+		}
+		break;
+
```
done.

fuyuntsuo commented:
> > Despite the changes above, I also moved a struct (L32:igo_nr.c) and a macro (L30:igo_nr.c) from `igo_nr.h` to `igo_nr.c` because they are intended to be called inside `igo_nr.c` only.
> 
> Thanks for the update @fuyuntsuo. Just a few more minor comments

NP. Thanks for reviewing the PR as well. All the issues have been addressed. Please review again. Thanks!

lgirdwood commented:
@fuyuntsuo please fix the comments above in a new PR - we can merge this fix PR for v1.8-rc2

fuyuntsuo commented:
> @fuyuntsuo please fix the comments above in a new PR - we can merge this fix PR for v1.8-rc2

No problem. We will address the remaining issues in the new PR. Thanks!

fuyuntsuo commented:
> > @fuyuntsuo please fix the comments above in a new PR - we can merge this fix PR for v1.8-rc2
> 
> No problem. We will address the remaining issues in the new PR. Thanks!

The new PR is #4152.
